pkg/migrations/registry.go

```diff
@@ -0,0 +1,161 @@
+package migrations
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/replicatedhq/embedded-cluster-operator/pkg/registry"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// registryData copies data from the disk (/var/lib/embedded-cluster/registry) to the seaweedfs s3 store.
+// if it fails, it will scale the registry deployment back to 1.
+// if it succeeds, it will create a secret used to indicate success to the operator.
+func registryData() error {
+	// if the migration fails, we need to scale the registry back to 1
+	success := false
+	defer func() {
+		if !success {
+			err := registryScale(1)
+			if err != nil {
+				fmt.Printf("Failed to scale registry back to 1 replica: %v\n", err)
+			}
+		}
+	}()
+	err := registryScale(0)
+	if err != nil {
+		return fmt.Errorf("failed to scale registry to 0 replicas before uploading data: %w", err)
+	}
+
+	fmt.Printf("Connecting to s3\n")
+	creds := credentials.NewStaticCredentialsProvider(os.Getenv("s3AccessKey"), os.Getenv("s3SecretKey"), "")
+	conf, err := config.LoadDefaultConfig(context.TODO(),
+		config.WithCredentialsProvider(creds),
+		config.WithRegion("us-east-1"),
+		config.WithEndpointResolverWithOptions(
+			aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
+				return aws.Endpoint{URL: "http://seaweedfs-s3.seaweedfs:8333"}, nil
+			})),
+	)
+	if err != nil {
+		return fmt.Errorf("load aws config: %w", err)
+	}
+
+	s3Client := s3.NewFromConfig(conf, func(o *s3.Options) {
+		o.UsePathStyle = true
+	})
+	registryStr := "registry"
+	_, err = s3Client.CreateBucket(context.TODO(), &s3.CreateBucketInput{
+		Bucket: &registryStr,
+	})
+	if err != nil {
+		if !strings.Contains(err.Error(), "BucketAlreadyExists") {
+			return fmt.Errorf("create bucket: %w", err)
+		}
+	}
+
+	fmt.Printf("Running registry data migration\n")
+	err = filepath.Walk("/var/lib/embedded-cluster/registry", func(path string, info os.FileInfo, err error) error {
```

Comment:
Is there a Go version of awscli sync command that the aws sdk Go package offers? Wondering if that sync command handles weird edge cases that could be missed here.

Reply:
ooof
I suspect I'm going to need to handle already-existing files, aren't I

Reply:
That, and hidden files, weird permissions, etc...

Reply:
At present it does handle already existing files, by overwriting them
I think this is fine here

It also doesn't have handling for odd files, but registry doesn't seem to create those

Reply:
Sounds good